### PR TITLE
UI:Updates the access and refresh token to keep user logged in

### PR DESCRIPTION
**Pull request by pratap0007**
This patch updates the access and refresh token before
previous respective token expiry time to keep user logged in

Signed-off-by: Shiv Verma <shverma@redhat.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

### CODE REVIEW

File: ui/src/containers/UserProfile/index.tsx [hunk 1]
```diff
@@ -15,16 +15,48 @@ import './UserProfile.css';
...
+
+      // setInterval for refresh the access token before 20 seconds of expiry time
+      const id2 = window.setInterval(() => {
+        user.updateAccessToken();
+      }, accessTokenInterval - 2000);
+      setIntervalId2(id2);
+    };
+
+    triggerInterval();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (vinamra28) [inline ui/src/containers/UserProfile/index.tsx:40]: instead of this can we add the plugin of `react-hooks` in `.eslintrc.json`?

File: ui/src/api/index.ts [hunk 2]
```diff
@@ -133,4 +136,36 @@ export class Hub implements Api {
       return err;
     }
   }
+
+  async getRefreshToken(refreshToken: any) {
```
Comment (vinamra28) [inline ui/src/api/index.ts:140]:
```suggestion
  async getRefreshToken(refreshToken: string) {
```

**Review by vinamra28 — CHANGES_REQUESTED** [review]

File: ui/src/containers/UserProfile/index.tsx [hunk 3]
```diff
@@ -15,16 +15,48 @@ import './UserProfile.css';
 const UserProfile: React.FC = () => {
   const { user } = useMst();
 
+  const [intervalId1, setIntervalId1] = useState<number>();
+  const [intervalId2, setIntervalId2] = useState<number>();
```
Comment (vinamra28) [inline ui/src/containers/UserProfile/index.tsx:19]: can we improve the naming over here as it's bit confusing :sweat_smile:

File: ui/src/containers/UserProfile/index.tsx [hunk 4]
```diff
@@ -15,16 +15,48 @@ import './UserProfile.css';
...
+
+  useEffect(() => {
+    const triggerInterval = () => {
+      const accessTokenInterval = user.accessTokenInfo.expiresAt * 1000 - new Date().getTime();
+      const refreshTokenInterval = user.refreshTokenInfo.expiresAt * 1000 - new Date().getTime();
+
+      // setInterval for refresh the refresh token before 20 seconds of expiry time
+      const id1 = window.setInterval(() => {
```
Comment (vinamra28) [inline ui/src/containers/UserProfile/index.tsx:27]: same as above

File: ui/src/containers/UserProfile/index.tsx [hunk 5]
```diff
@@ -15,16 +15,48 @@ import './UserProfile.css';
...
+        user.updateRefreshToken();
+      }, refreshTokenInterval - 2000);
+      setIntervalId1(id1);
+
+      // setInterval for refresh the access token before 20 seconds of expiry time
+      const id2 = window.setInterval(() => {
```
Comment (vinamra28) [inline ui/src/containers/UserProfile/index.tsx:33]: same as above

File: ui/src/store/auth.test.ts [hunk 6]
```diff
@@ -141,4 +141,29 @@ describe('Store functions', () => {
...
+    store.updateRefreshToken();
+    expect(store.isLoading).toBe(true);
+    when(
+      () => !store.isLoading,
+      () => {
+        setTimeout(() => {
+          expect(store.isAuthenticated).toBe(true);
+
+          expect(store.accessTokenInfo.expiresAt).toBe(1611888148);
+          expect(store.refreshTokenInfo.expiresAt).toBe(1611888989);
```
Comment (vinamra28) [inline ui/src/store/auth.test.ts:162]: can we add a comment over here so that it's clear how many min or hrs or days it represents?

File: ui/.eslintrc.json [hunk 7]
```diff
@@ -6,7 +6,7 @@
     "prettier/@typescript-eslint",
     "plugin:prettier/recommended"
   ],
-  "plugins": ["react", "@typescript-eslint", "prettier"],
+  "plugins": ["react","react-hooks", "@typescript-eslint", "prettier"],
```
Comment (PuneetPunamiya) [inline ui/.eslintrc.json:9]: Is this change related to this pr ??

File: ui/src/containers/Header/Header.test.tsx [hunk 8]
```diff
@@ -12,6 +12,14 @@ const TESTDATA_DIR = `src/store/testdata`;
 const api = new FakeHub(TESTDATA_DIR);
 const { Provider, root } = createProviderAndStore(api);
 
+beforeEach(() => {
```
Comment (PuneetPunamiya) [inline ui/src/containers/Header/Header.test.tsx:15]: Can we please add some description in the comment to understand what `beforeEach`  and `afterEach` actually does and why do we need them ?

File: ui/src/containers/Header/index.tsx [hunk 9]
```diff
@@ -28,7 +28,7 @@ const Header: React.FC = observer(() => {
           <Search />
         </GridItem>
       </Grid>
-      {user.isAuthenticated ? (
+      {user.isAuthenticated && user.refreshTokenInfo.expiresAt * 1000 > global.Date.now() ? (
```
Comment (PuneetPunamiya) [inline ui/src/containers/Header/index.tsx:31]: What wil lhappen if `user.refreshTokenInfo.expiresAt * 1000 `< global.Date.now()`

File: ui/src/containers/UserProfile/index.tsx [hunk 10]
```diff
@@ -15,16 +15,48 @@ import './UserProfile.css';
...
+
+      // setInterval for refresh the access token before 20 seconds of expiry time
+      const tempAccessId = window.setInterval(() => {
+        user.updateAccessToken();
+      }, accessTokenInterval - 2000);
+      setAccessId(tempAccessId);
+    };
+
+    triggerInterval();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (PuneetPunamiya) [inline ui/src/containers/UserProfile/index.tsx:40]: Can we please remove this ??

File: ui/src/containers/Header/index.tsx (hunk 9 above)
Comment (pratap0007) [reply ui/src/containers/Header/index.tsx:31]: user will be logged out

File: ui/.eslintrc.json (hunk 7 above)
Comment (pratap0007) [reply ui/.eslintrc.json:9]: yes

File: ui/src/containers/Header/index.tsx (hunk 9 above)
Comment (PuneetPunamiya) [reply ui/src/containers/Header/index.tsx:31]: So new refresh token api won't be called ??

File: ui/src/containers/UserProfile/index.tsx [hunk 11]
```diff
@@ -15,16 +15,48 @@ import './UserProfile.css';
...
+  }, []);
+
   const [isModalOpen, setIsModalOpen] = useState(false);
   const [isOpen, set] = useState(false);
 
+  const hubLogout = () => {
```
Comment (PuneetPunamiya) [inline ui/src/containers/UserProfile/index.tsx:46]: Should we move this function to store :thinking: ?

File: ui/src/containers/Header/index.tsx (hunk 9 above)
Comment (pratap0007) [reply ui/src/containers/Header/index.tsx:31]: yes

File: ui/src/containers/UserProfile/index.tsx [hunk 12]
```diff
@@ -15,16 +15,48 @@ import './UserProfile.css';
 const UserProfile: React.FC = () => {
   const { user } = useMst();
 
+  const [refreshId, setRefreshId] = useState<number>();
```
Comment (PuneetPunamiya) [inline ui/src/containers/UserProfile/index.tsx:18]: Here what will the initial value of `refreshId`

File: ui/src/containers/UserProfile/index.tsx (hunk 11 above)
Comment (pratap0007) [reply ui/src/containers/UserProfile/index.tsx:46]: `localStore.clear()` was not working in the store so I kept here

File: ui/src/containers/UserProfile/index.tsx (hunk 12 above)
Comment (pratap0007) [reply ui/src/containers/UserProfile/index.tsx:18]: undefined

Comment (PuneetPunamiya) [reply ui/src/containers/UserProfile/index.tsx]: Then i think it is not right to have `undefined` values as by default

File: ui/src/containers/UserProfile/index.tsx (hunk 11 above)
Comment (PuneetPunamiya) [reply ui/src/containers/UserProfile/index.tsx:47]: Other than `localstorage` can be just kept other can me moved right ???

Comment (pratap0007) [reply ui/src/containers/UserProfile/index.tsx:47]: yeah that's already in the store

File: ui/src/containers/UserProfile/index.tsx (hunk 12 above)
Comment (pratap0007) [reply ui/src/containers/UserProfile/index.tsx]: done !

File: ui/src/containers/Header/Header.test.tsx [hunk 13]
```diff
@@ -12,6 +12,16 @@ const TESTDATA_DIR = `src/store/testdata`;
 const api = new FakeHub(TESTDATA_DIR);
 const { Provider, root } = createProviderAndStore(api);
 
+// Assign a dummy date to global Date inorder to avoid test failue
+beforeEach(() => {
+  global.Date.now = jest.fn(() => new Date('2021-01-01T10:20:30Z').getTime());
```
Comment (PuneetPunamiya) [inline ui/src/containers/Header/Header.test.tsx:17]: I think we should create a function for this in `common` directory since it is used at many places

File: ui/src/containers/Header/Header.test.tsx [hunk 14]
```diff
@@ -12,6 +12,16 @@ const TESTDATA_DIR = `src/store/testdata`;
...
+});
+
+// Assign current Date once test is over
+afterEach(() => {
```
Comment (PuneetPunamiya) [inline ui/src/containers/Header/Header.test.tsx:21]: Same as above

File: ui/src/containers/UserProfile/index.tsx [hunk 15]
```diff
@@ -15,16 +15,49 @@ import './UserProfile.css';
...
+  const [refreshId, setRefreshId] = useState<number>(0);
+  const [accessId, setAccessId] = useState<number>(0);
+
+  const triggerInterval = useCallback(() => {
+    const accessTokenInterval = user.accessTokenInfo.expiresAt * 1000 - new Date().getTime();
+    const refreshTokenInterval = user.refreshTokenInfo.expiresAt * 1000 - new Date().getTime();
+
+    // To get a new refresh token
+    // Update the refresh token before 10 seconds of current refresh token's expiry time
+    const tempRefreshId = window.setInterval(() => {
```
Comment (PuneetPunamiya) [inline ui/src/containers/UserProfile/index.tsx:27]: I think we should create an action in store to handle this case or may be we can create a function in utils where we pass the tokens and get the result

File: ui/src/containers/Header/Header.test.tsx (hunk 13 above)
Comment (vinamra28) [reply ui/src/containers/Header/Header.test.tsx:17]: this is a hook I guess....so not sure whether we can place it in commons directory :thinking:

File: ui/src/containers/UserProfile/index.tsx (hunk 15 above)
Comment (pratap0007) [reply ui/src/containers/UserProfile/index.tsx:27]: with action I had tried once but was not working

File: ui/src/containers/Header/Header.test.tsx (hunk 13 above)
Comment (pratap0007) [reply ui/src/containers/Header/Header.test.tsx:17]: okay will check it once :+1:

File: ui/src/store/auth.ts [hunk 16]
```diff
@@ -160,6 +159,40 @@ export const AuthStore = types
...
+        const { api } = self;
+
+        const refresh = yield api.getRefreshToken(self.refreshTokenInfo.token);
+        const newRefreshToken = refresh.data;
+        self.addRefreshTokenInfo(newRefreshToken.refresh);
+      } catch (err) {
+        const error: IError = {
+          status: err.status,
+          serverMessage: titleCase(err.data.message),
+          customMessage: ''
```
Comment (vinamra28) [inline ui/src/store/auth.ts:175]: don't we need any custom message over here?

File: ui/src/store/auth.ts [hunk 17]
```diff
@@ -160,6 +159,40 @@ export const AuthStore = types
...
+        const { api } = self;
+
+        const access = yield api.getAccessToken(self.refreshTokenInfo.token);
+        const newAccessToken = access.data;
+        self.addAccessTokenInfo(newAccessToken.access);
+      } catch (err) {
+        const error: IError = {
+          status: err.status,
+          serverMessage: titleCase(err.data.message),
+          customMessage: ''
```
Comment (vinamra28) [inline ui/src/store/auth.ts:192]: same as above

**Comment by SM43** [comment]
lgtm
/thank you

File: ui/src/api/index.ts [hunk 18]
```diff
@@ -133,4 +136,36 @@ export class Hub implements Api {
...
+    }
+  }
+
+  async getAccessToken(refreshToken: string) {
+    try {
+      return axios({
+        method: 'post',
+        url: `${API_URL}/user/refresh/accesstoken`,
+        headers: {
+          Authorization: `Bearer ${refreshToken}`
```
Comment (vinamra28) [inline ui/src/api/index.ts:162]: can we merge these two functions and make a common function?
